packages/ra-ui-materialui/src/list/DatagridRow.js

```diff
@@ -62,33 +62,36 @@ class DatagridRow extends Component {
 
         if (!rowClick) return;
 
-        if (typeof rowClick === 'function') {
-            const path = await rowClick(id, basePath, record);
-            this.handleRedirection(path, event);
-            return;
-        }
+        const path =
+            typeof rowClick === 'function'
+                ? await rowClick(id, basePath, record)
+                : rowClick;
 
-        this.handleRedirection(rowClick, event);
+        this.handleRedirection(path, event);
     };
 
     handleRedirection = (path, event) => {
         const { basePath, id, push } = this.props;
 
-        if (path === 'edit') {
-            push(linkToRecord(basePath, id));
-            return;
-        }
-        if (path === 'show') {
-            push(linkToRecord(basePath, id, 'show'));
-            return;
-        }
-        if (path === 'expand') {
-            this.handleToggleExpanded(event);
-            return;
-        }
         if (!path) return;
 
-        push(path);
+        switch (path) {
+            case 'edit':
+                push(linkToRecord(basePath, id));
+                return;
+            case 'show':
+                push(linkToRecord(basePath, id, 'show'));
+                return;
+            case 'expand':
+                this.handleToggleExpanded(event);
+                return;
+            case 'toggleSelection':
+                this.handleToggle(event);
+                return;
+            default:
+                push(path);
```

Comment:
If path is empty, the previous code didn't push. You're introducing  breaking change here.

Reply:
No because there is a condition just before the `switch` :)

Reply:
However I agree that it would be clearer with this condition in the `default`